app/components/machine/driver-hwclouds/component.js

```diff
@@ -0,0 +1,377 @@
+      var kService =  AWS.util.crypto.hmac(kRegion, sn, 'buffer');
+      var kSigning =  AWS.util.crypto.hmac(kService, 'sdk_request', 'buffer');
+      return kSigning;
+    }
+
+    function signStringToSign(stringToSign, signingKey) {
+      return AWS.util.crypto.hmac(signingKey, stringToSign, 'hex');
+    }
+
+    function credentialScope (time, rname, sname) {
+      return `${time}/${rname}/${sname}/sdk_request`;
+    }
+
+    function buildCanonicalHeaders(requestHeaders) {
+      var headers =  [];
+      var parts =    [];
+
+      Object.keys(requestHeaders).forEach((key) => {
+        headers.push([key, requestHeaders[key]]);
+      });
+
+      headers.sort(function (a, b) {
+        return a[0].toLowerCase() < b[0].toLowerCase() ? -1 : 1;
+      });
+
+      headers.forEach((item) => {
+        var key = item[0].toLowerCase();
+        parts.push(`${key}:${item[1].toString().replace(/\s+/g, ' ').replace(/^\s+|\s+$/g, '')}`);
+      });
+
+      return `${parts.join('\n')}\n`;
+    }
+
+    function buildCanonicalQS(qs) {
+      var qsParts = qs.split('&');
+
+      qsParts.sort(function (a, b) {
+        return a[0].toLowerCase() < b[0].toLowerCase() ? -1 : 1;
+      });
+
+      qsParts.forEach((qString) => {
+        qString.replace(/\+/g, '%20');
+      });
+
+      return qsParts.join('&');
+    }
+
+    function buildCanonicalRequest(request) {
+      var parts = [];
+
+      parts.push(request.method);
+      parts.push(encodeURI(request.urlToEncode));
+      parts.push(buildCanonicalQS(request.queryString));
+      parts.push(buildCanonicalHeaders(request.headers));
+      parts.push('content-type;host;x-sdk-date');
+      parts.push(AWS.util.crypto.sha256(request.body, 'hex'));
+
+      return parts.join('\n');
+    }
+
+    function stringToSign(canonicalRequest, credentialScope, t) {
+      var parts = [];
+
+      parts.push('SDK-HMAC-SHA256');
+      parts.push(t);
+      parts.push(credentialScope);
+      parts.push(canonicalRequest);
+
+      return parts.join('\n');
+    }
+
+    return addAuthorization(credentials);
+  },
+
+  buildAuth: function(params) {
+    var out = [];
+
+    out.push(`${params.SignatureMethod} Credential=${this.get('model.hwcloudsConfig.accessKeyId')}/${params.ShortTimestamp}/eu-de/ecs/sdk_request`);
+    out.push('SignedHeaders=content-type;host;x-sdk-date');
+    out.push(`Signature=${params.Signature}`);
+
+    return out.join(', ');
+  },
+
+  getSecurityGroups: function() {
+    var time = moment.utc();
+    var queryParams = `vpc_id=${this.get('model.hwcloudsConfig.vpcId')}`;
+    var signature = this.signAuth({
+      accessKeyId: this.get('model.hwcloudsConfig.accessKeyId'),
+      secretAccessKey: this.get('model.hwcloudsConfig.accessKeySecret')
+    }, this.describeRequest(this.getURL('ecs', 'v1', this.get('model.hwcloudsConfig.tenantId'), 'security-groups', queryParams), 'GET', `${time.format('YYYYMMDDTHHmmss')}Z`, ''), time);
+
+    return this.apiRequest(`${signature.url}`, {
```

Comment:
Can the time generation and signing and adding the header go into `apiRequest()` instead of being duplicated in every API call